### PR TITLE
Task2

### PR DESCRIPTION
**Pull request by LesterOper**
Finished task

### CODE REVIEW

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 1]
```diff
@@ -1,43 +1,61 @@
...
     
     @Override
-    public List<Product> findProducts() {
-        return result.stream().
+    public List<Product> findProducts(String query, SortField field, SortPrice price) {
+            return result.stream().
                 filter(x-> x.getPrice()!=null && x.getStock()>0).
+                filter(x-> query==null || x.getDescription().contains(query)).
+                sorted(getComparator(field, price)).
                 collect(Collectors.toCollection(ArrayList::new));
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Here and in other places: streams formatting: https://marcin-chwedczuk.github.io/java-streams-best-practices

File: src/main/webapp/WEB-INF/pages/productList.jsp [hunk 2]
```diff
@@ -8,20 +8,28 @@
...
+        </td>
       </tr>
     </thead>
     <c:forEach var="product" items="${products}">
       <tr>
         <td>
           <img class="product-tile" src="https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/${product.imageUrl}">
         </td>
-        <td>${product.description}</td>
+        <td> <a href = "products/description?prod=${product.id}">${product.description}</a></td>
```
Comment (SergeiStarikov) [inline src/main/webapp/WEB-INF/pages/productList.jsp]: Formatting improves readability:
```
<td>
    <a>...</a>
</td>
```

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 3]
```diff
@@ -11,14 +11,28 @@
...
-        this.product = new ArrayListProductDao();
+        this.product = ArrayListProductDao.getInstance();
     }
     
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        request.setAttribute("products",product.findProducts());
+        SortPrice price = SortPrice.no;
+        SortField field = SortField.no;
+        if(request.getParameter("sort")!=null){
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java:25]: Consider moving repeating strings to constants.

File: src/main/java/com/es/phoneshop/web/ProductDetailsServlet.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+
+public class ProductDetailsServlet extends HttpServlet{
+    
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException{
+        request.setAttribute("products",product.getProduct(Long.parseLong(request.getParameter("prod"))));
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/web/ProductDetailsServlet.java]: The line is too long. Consider splitting into multiple.

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 5]
```diff
@@ -53,7 +71,7 @@ public synchronized void save(Product product) {
 
     @Override
     public synchronized void delete(Long id) {
-        result.stream().filter(x->x.getId().equals(id)).
+        result.stream().filter(x-> x.getId()!=null||x.getId().equals(id)).
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java:88]: x.getId()!=null && x.getId().equals(id) ?

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 6]
```diff
@@ -1,43 +1,61 @@
...
                 orElseThrow(IllegalArgumentException::new);
     }
     
     @Override
-    public List<Product> findProducts() {
-        return result.stream().
+    public List<Product> findProducts(String query, SortField field, SortPrice price) {
+            return result.stream().
                 filter(x-> x.getPrice()!=null && x.getStock()>0).
+                filter(x-> query==null || x.getDescription().contains(query)).
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: More functionality expected here: the task is "The search must work using OR clause. The example above means: search all products having description containing “Samsung” or “S” or “II”. Products having all search terms must come first in the search result."

File: src/main/webapp/WEB-INF/web.xml [hunk 7]
```diff
@@ -14,5 +16,21 @@
...
+      <listener-class>
+          com.es.phoneshop.listener.ProductContextListener
+      </listener-class>
+  </listener>
+  
+  <servlet>
+    <servlet-name>productDetails</servlet-name>
+    <servlet-class>com.es.phoneshop.web.ProductDetailsServlet</servlet-class>
+  </servlet>
+
```
Comment (SergeiStarikov) [inline src/main/webapp/WEB-INF/web.xml:30]: <error-page> here?

Task "Display 404 page with “Product with code xxx not found” message if a product can’t be found."

File: src/main/webapp/WEB-INF/pages/productList.jsp [hunk 8]
```diff
@@ -8,20 +8,28 @@
...
         <td class="price">
```
Comment (SergeiStarikov) [inline src/main/webapp/WEB-INF/pages/productList.jsp:36]: Price history popup here?

Task "Clicking product price shall open a popup with price change history."

Comment (LesterOper) [reply src/main/webapp/WEB-INF/pages/productList.jsp:36]: This task is selfcheck and I thought it wasn't necessary to do by the first check

File: src/main/webapp/WEB-INF/web.xml (hunk 7 above)
Comment (LesterOper) [reply src/main/webapp/WEB-INF/web.xml:30]: oooo, I forgot about it

File: src/main/java/com/es/phoneshop/exception/ProductNotFound.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
+
+package com.es.phoneshop.exception;
+
+
+public class ProductNotFound extends RuntimeException{
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/exception/ProductNotFound.java:5]: ProductNotFoundException

**Pull request merged**